src/twisted/python/systemd.py

```diff
@@ -68,26 +70,70 @@ def fromEnvironment(
         if start is None:
             start = cls._START
 
-        descriptors: List[int] = []
-
-        try:
-            pid = int(environ["LISTEN_PID"])
-        except (KeyError, ValueError):
-            pass
+        if str(getpid()) == environ.get("LISTEN_PID"):
+            descriptors: List[int] = _parseDescriptors(start, environ)
+            names: Sequence[str] = _parseNames(environ)
         else:
-            if pid == getpid():
-                try:
-                    count = int(environ["LISTEN_FDS"])
-                except (KeyError, ValueError):
-                    pass
-                else:
-                    descriptors = list(range(start, start + count))
-                    del environ["LISTEN_PID"], environ["LISTEN_FDS"]
-
-        return cls(descriptors)
-
-    def inheritedDescriptors(self) -> Iterable[int]:
+            descriptors = []
+            names = ()
+
+        # They may both be missing (consistent with not running under systemd
+        # at all) or they may both be present (consistent with running under
+        # systemd 227 or newer).  It is not allowed for only one to be present
+        # or for the values to disagree with each other.
+        if len(names) != len(descriptors):
+            return cls([], ())
+
+        return cls(descriptors, names)
+
+    def inheritedDescriptors(self) -> List[int]:
         """
         @return: The configured descriptors.
         """
         return list(self._descriptors)
+
+    def inheritedNamedDescriptors(self) -> Dict[str, int]:
+        """
+        @return: A mapping from the names of configured descriptors to
+            their integer values.
+        """
+        return dict(zip(self._names, self._descriptors))
+
+
+def _parseDescriptors(start: int, environ: Mapping[str, str]) -> List[int]:
+    """
+    Parse the I{LISTEN_FDS} environment variable supplied by systemd.
+
+    @param start: systemd provides only a count of the number of descriptors
+        that have been inherited.  This is the integer value of the first
+        inherited descriptor.  Subsequent inherited descriptors are numbered
+        counting up from here.  See L{ListenFDs._START}.
+
+    @param environ: The environment variable mapping in which to look for the
+        value to parse.
+
+    @return: The integer values of the inherited file descriptors, in order.
+    """
+    try:
+        count = int(environ["LISTEN_FDS"])
+    except (KeyError, ValueError):
+        return []
+    else:
+        descriptors = list(range(start, start + count))
+        del environ["LISTEN_PID"], environ["LISTEN_FDS"]
```

Comment:
I left a comment in the test case.
It is not clear why we are doing this.

I know this is already in `trunk` but looks like a strange side effect.

```suggestion
        # We do THIS, because of THAT.
        del environ["LISTEN_PID"], environ["LISTEN_FDS"]
```